src/mlpack/methods/reinforcement_learning/q_learning.hpp

```diff
@@ -0,0 +1,170 @@
+ *               Alex Graves and
+ *               Ioannis Antonoglou and
+ *               Daan Wierstra and
+ *               Martin A. Riedmiller},
+ *  title     = {Playing Atari with Deep Reinforcement Learning},
+ *  journal   = {CoRR},
+ *  year      = {2013},
+ *  url       = {http://arxiv.org/abs/1312.5602}
+ * }
+ * @endcode
+ *
+ * @tparam EnvironmentType The environment of the reinforcement learning task.
+ * @tparam NetworkType The network to compute action value.
+ * @tparam OptimizerType The optimizer to train the network.
+ * @tparam PolicyType Behavior policy of the agent.
+ * @tparam ReplayType Experience replay method.
+ */
+template <
+  typename EnvironmentType,
+  typename NetworkType,
+  typename OptimizerType,
+  typename PolicyType,
+  typename ReplayType = RandomReplay<EnvironmentType>
+>
+class QLearning
+{
+ public:
+  //! Convenient typedef for state.
+  using StateType = typename EnvironmentType::State;
+
+  //! Convenient typedef for action.
+  using ActionType = typename EnvironmentType::Action;
+
+  /**
+   * Create the QLearning object with given settings.
+   *
+   * @param network The network to compute action value.
+   * @param optimizer The optimizer to train the network.
+   * @param discount Discount for future return.
+   * @param policy Behavior policy of the agent.
+   * @param replayMethod Experience replay method.
+   * @param targetNetworkSyncInterval Interval (steps) to sync the target network.
+   * @param explorationSteps Steps before starting to learn.
+   * @param doubleQLearning Whether to use double Q-Learning.
+   * @param stepLimit Maximum steps in each episode, 0 means no limit.
+   * @param environment Reinforcement learning task.
+   */
+  QLearning(NetworkType& network,
+            OptimizerType& optimizer,
+            double discount,
+            PolicyType policy,
+            ReplayType replayMethod,
+            size_t targetNetworkSyncInterval,
+            size_t explorationSteps,
+            bool doubleQLearning = false,
+            size_t stepLimit = 0,
+            EnvironmentType environment = EnvironmentType());
+
+  /**
+   * Execute a step in an episode.
+   * @return Reward for the step.
+   */
+  double Step();
+
+  /**
+   * Execute an episode.
+   * @return Return of the episode.
+   */
+  double Episode();
+
+  /**
+   * @return Total steps from beginning.
+   */
+  const size_t& TotalSteps() const { return totalSteps; }
+
+  //! Modify the training mode / test mode indicator.
+  bool& Deterministic() { return deterministic; }
+
+  //! Get the indicator of training mode / test mode.
+  const bool& Deterministic() const { return deterministic; }
+
+ private:
+  /**
+   * Select the best action based on given action value.
+   * @param actionValues Action values.
+   * @return Selected actions.
+   */
+  arma::icolvec BestAction(const arma::mat& actionValues);
+
+  //! Reference of the learning network.
+  NetworkType& learningNetwork;
```

Comment:
For solution 2 we still have to provide the existing constructor and also the existing ``Optimize(arma::mat& iterate, FunctionType& function);`` to be backward compatible at least for mlpack < 3.0, so that's this. Also despite the refactoring that we have to do, I can't think of any problem that comes with the solution. I would also prefer solution 2 over 3, not just because you don't have to think about rebinding the function in certain cases. But the big point is backward compatibility. Eager to here @rcurtin on this point.